store/tikv/range_task.go

```diff
@@ -0,0 +1,227 @@
+// limitations under the License.
+
+package tikv
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const (
+	rangeTaskDefaultStatLogInterval = time.Minute * 10
+)
+
+// RangeTaskRunner splits a range into many ranges to process concurrently, and convenient to send requests to all
+// regions in the range. Because of merging and splitting, it's possible that multiple requests for disjoint ranges are
+// sent to the same region.
+type RangeTaskRunner struct {
+	name            string
+	store           Storage
+	concurrency     int
+	handler         RangeTaskHandler
+	statLogInterval time.Duration
+
+	completedRegions int32
+}
+
+// RangeTaskHandler is the type of functions that processes a task of a key range.
+// The function can update completedRegions via atomic operations.
+type RangeTaskHandler = func(ctx context.Context, r kv.KeyRange, completedRegions *int32) error
+
+// NewRangeTaskRunner creates a RangeTaskRunner.
+//
+// `requestCreator` is the function used to create RPC request according to the given range.
+// `responseHandler` is the function to process responses of errors. If `responseHandler` returns error, the whole job
+// will be canceled.
+func NewRangeTaskRunner(
+	name string,
+	store Storage,
+	concurrency int,
+	handler RangeTaskHandler,
+) *RangeTaskRunner {
+	return &RangeTaskRunner{
+		name:            name,
+		store:           store,
+		concurrency:     concurrency,
+		handler:         handler,
+		statLogInterval: rangeTaskDefaultStatLogInterval,
+	}
+}
+
+// SetStatLogInterval sets the time interval to log the stats.
+func (s *RangeTaskRunner) SetStatLogInterval(interval time.Duration) {
+	s.statLogInterval = interval
+}
+
+// RunOnRange runs the task on the given range.
+// Empty startKey or endKey means unbounded.
+func (s *RangeTaskRunner) RunOnRange(ctx context.Context, startKey []byte, endKey []byte) error {
+	s.completedRegions = 0
+
+	if len(endKey) != 0 && bytes.Compare(startKey, endKey) >= 0 {
+		logutil.Logger(ctx).Info("empty range task executed. ignored",
+			zap.String("name", s.name),
+			zap.Binary("startKey", startKey),
+			zap.Binary("endKey", endKey))
+		return nil
+	}
+
+	logutil.Logger(ctx).Info("range task started",
+		zap.String("name", s.name),
+		zap.Binary("startKey", startKey),
+		zap.Binary("endKey", endKey))
+
+	// Periodically log the progress
+	statLogTicker := time.NewTicker(s.statLogInterval)
+
+	ctx, cancel := context.WithCancel(ctx)
+	taskCh := make(chan *kv.KeyRange, s.concurrency)
+	var wg sync.WaitGroup
+
+	// Create workers that concurrently process the whole range.
+	for i := 0; i < s.concurrency; i++ {
+		w := s.createWorker(taskCh, &wg)
+		wg.Add(1)
+		go w.run(ctx, cancel)
+	}
+
+	startTime := time.Now()
+
+	defer func() {
+		close(taskCh)
+		statLogTicker.Stop()
+		wg.Wait()
+		cancel()
```

Comment:
What if some ranges report error after `close(taskCh)`? Once you get into `defer`, the function will return `nil`.